### PR TITLE
Document common pitfalls

### LINKED ISSUE #404: Document common pitfalls
There are some points that continue to confuse people (especially as it relates to our async behavior).

One in particular is using closure scope variables to try and pass data between `beforeEach` and individual tests (see #402). 

I think we need to create some documentation that addresses each of these types misconceptions so we can quickly link to them. They can be in the `readme` if they are concise, but maybe separate documents if they require lengthy explanations.

If anyone can think of other common downfalls / misconceptions / misuses - go ahead and add them to this discussion.

### PR DESCRIPTION
**Pull request by sotojuan**
Fixes #404.

I added @novemberborn's Docker idea.

I'm looking for more pitfalls and what I can write about them. Here are some from the issue:

> Using global variables and forgetting that tests run concurrently

Does saying "AVA run test files in different processes so there's no shared global state..." suffice?

> `console.log` not logging where the test result is printed

What's the status on this with the closed issue (#392)?

Any other ideas welcome, this can be WIP until we have a good amount.

### CODE REVIEW

File: docs/common-pitfalls.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# Common Pitfalls
+
+## AVA in Docker
+
+If you run AVA in Docker as part of your CI, you need to fix the appropriate environment variables. Specifically, adding `-e CI=true` in the `docker exec` command. See [https://github.com/avajs/ava/issues/751](#751).
+
+AVA uses [is-ci](https://github.com/watson/is-ci) to decide if it's in a CI environment or not using [these](https://github.com/watson/is-ci/blob/master/index.js) variables.
```
Comment (sindresorhus) [inline docs/common-pitfalls.md]: > AVA uses is-ci to decide whether it's in a CI environment using these environment variables.

**Comment by sindresorhus** [comment]
Good start.

Needs to be added to the recipe section in the readme.

File: docs/common-pitfalls.md [hunk 2]
```diff
@@ -5,3 +5,9 @@
 If you run AVA in Docker as part of your CI, you need to fix the appropriate environment variables. Specifically, adding `-e CI=true` in the `docker exec` command. See [https://github.com/avajs/ava/issues/751](#751).
 
 AVA uses [is-ci](https://github.com/watson/is-ci) to decide if it's in a CI environment or not using [these](https://github.com/watson/is-ci/blob/master/index.js) variables.
+
+## AVA and connected client limits
+
+You may be using a service that only allows a limited number of concurrent connections. For example, many database-as-a-service businesses offer a free plan with a limit on how many clients can be using it at the same time. AVA can hit those limits and cause the tests to hang because it runs multiple processes, each of which is a different client.
```
Comment (sindresorhus) [inline docs/common-pitfalls.md]: I think we should be clear that it's those services that are being bad and not AVA. They shouldn't hang AVA just because it's concurrency greedy.

Comment (jamestalmage) [inline docs/common-pitfalls.md]: > Well written clients / services should emit errors if concurrency limits are exceeded, instead of transparently hanging execution. If they don't, you should consider submitting an issue request asking that they do.

Comment (sindresorhus) [inline docs/common-pitfalls.md]: @jamestalmage No, they should just handle it with throttling, but very few do...

Comment (jamestalmage) [inline docs/common-pitfalls.md]: IDK, the whole point seems a little unnecessary. Services are going to do what they do, we are just explaining the reality of how to make AVA work with services that behave this way.

Comment (sotojuan) [inline docs/common-pitfalls.md]: 👍

**Comment by sotojuan** [comment]
Updated.

**Comment by novemberborn** [comment]
Maybe add a call to action for people reading this document and thinking "but my problem isn't listed here"?

**Comment by sindresorhus** [comment]
@sotojuan Interested in including something about https://github.com/avajs/ava/issues/404#issuecomment-226881829 here, or would you prefer doing that in a separate PR?

**Comment by sotojuan** [comment]
I'll add it! We can close the PR when we think we have enough to get started. Will write it up after I finish breakfast.

**Comment by sotojuan** [comment]
Updated.

File: docs/common-pitfalls.md [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+## AVA and connected client limits
+
+You may be using a service that only allows a limited number of concurrent connections. For example, many database-as-a-service businesses offer a free plan with a limit on how many clients can be using it at the same time. AVA can hit those limits as it runs multiple processes, but well-written services should emit an error or throttle in those cases. If the one you're using doesn't, the tests will hang.
+
+Use the `concurrency` flag to limit the number of processes ran. For example, if your service plan allows 5 clients, you should run AVA with `concurrency=5` or less.
+
+## Async operations
+
+You may be running an async operation inside a test and wondering why it's not finishing. If your async operation uses promises, you should be returning the promise:
```
Comment (novemberborn) [inline docs/common-pitfalls.md]: "you should return the promise"

File: docs/common-pitfalls.md [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  });
+});
+```
+
+If it uses callbacks, use [`test.cb`](https://github.com/avajs/ava#callback-support):
+
+```js
+test.cb(t => {
+  fetch((err, data) => {
+    t.is(data, 'bar');
```
Comment (novemberborn) [inline docs/common-pitfalls.md]: I wonder if we should advise promisifying the function. Assertion failures would throw, which depending on the async function's implementation may cause an uncaught exception.

Also you'd need `t.end()` in this example.

File: docs/common-pitfalls.md (hunk 1 above)
Comment (novemberborn) [inline docs/common-pitfalls.md]: Maybe link `these variables`? It makes the link more descriptive.

File: docs/common-pitfalls.md [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  });
+});
+```
+
+---
+
+Is your problem not listed here? Submit a pull request or comment on [this](https://github.com/avajs/ava/issues/404) issue.
```
Comment (novemberborn) [inline docs/common-pitfalls.md]: Same here linking "this issue".

File: docs/common-pitfalls.md (hunk 4 above)
Comment (sindresorhus) [inline docs/common-pitfalls.md]: Yes, we should recommend promisifying, but still show both solutions.

Comment (jamestalmage) [inline docs/common-pitfalls.md]: > Assertion failures would throw

Our assertions don't throw after being enhanced.

```js
test('foo', t => {
  t.is(1, 2);
  console.log('foo'); // this will get printed.
});
```

Comment (novemberborn) [inline docs/common-pitfalls.md]: OK, but the next assertion could try to access a property of a `null` value and it'll still crash.

**Comment by sotojuan** [comment]
🐔

**Comment by sindresorhus** [comment]
> Needs to be linked to from the readme.

**Comment by sotojuan** [comment]
Got it—put it in Support if that's a good place.

**Comment by novemberborn** [comment]
@sotojuan probably better under `Tips`.

**Comment by sotojuan** [comment]
Updated.

File: readme.md [hunk 6]
```diff
@@ -68,13 +68,13 @@ $ ava --init
...
 ```json
 {
-  "name": "awesome-package",
-  "scripts": {
-    "test": "ava"
-  },
-  "devDependencies": {
-    "ava": "^0.11.0"
-  }
+	"name": "awesome-package",
```
Comment (novemberborn) [inline readme.md]: Could you leave the whitespace as it was?

File: readme.md [hunk 7]
```diff
@@ -966,6 +966,10 @@ You can't use [`istanbul`](https://github.com/gotwarlost/istanbul) for code cove
 
 As of version `5.0.0` it uses source maps to report coverage for your actual code, regardless of transpilation. Make sure that the code you're testing includes an inline source map or references a source map file. If you use `babel-register` you can set the `sourceMaps` option in your Babel config to `inline`.
 
+### Common pitfalls
+
+We have a growing list of [common pitfalls](https://github.com/avajs/ava/blob/master/docs/common-pitfalls.md) you may experience while using AVA. If you encounter any issues you think are common, comment on [this issue](https://github.com/avajs/ava/issues/404).
```
Comment (novemberborn) [inline readme.md]: Other links are relative (see `Recipes` for instance).

File: readme.md (hunk 6 above)
Comment (sotojuan) [inline readme.md]: This was part of the merge conflict (this part was an old version from whenever I made my fork). Fixed when I merged readme from master.

File: readme.md [hunk 8]
```diff
@@ -175,27 +175,27 @@ All of the CLI options can be configured in the `ava` section of your `package.j
...
+			"!foo"
+		],
+		"concurrency": 5,
+		"failFast": true,
+		"tap": true,
+		"require": [
+			"babel-register"
+		],
+		"babel": "inherit"
+	}
```
Comment (sindresorhus) [inline readme.md]: Don't do unrelated changes.

File: readme.md [hunk 9]
```diff
@@ -968,6 +968,10 @@ You can't use [`istanbul`](https://github.com/gotwarlost/istanbul) for code cove
 
 As of version `5.0.0` it uses source maps to report coverage for your actual code, regardless of transpilation. Make sure that the code you're testing includes an inline source map or references a source map file. If you use `babel-register` you can set the `sourceMaps` option in your Babel config to `inline`.
 
+### Common pitfalls
+
+We have a growing list of [common pitfalls](docs/common-pitfalls.md) you may experience while using AVA. If you encounter any issues you think are common, comment on [this issue](https://github.com/avajs/ava/issues/404).
```
Comment (sindresorhus) [inline readme.md]: Kinda funny that the issue number is 404 :p

Comment (sindresorhus) [inline readme.md]: comment on => comment in

File: docs/common-pitfalls.md [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+
+AVA uses [is-ci](https://github.com/watson/is-ci) to decide if it's in a CI environment or not using [these variables](https://github.com/watson/is-ci/blob/master/index.js).
+
+## AVA and connected client limits
+
+You may be using a service that only allows a limited number of concurrent connections. For example, many database-as-a-service businesses offer a free plan with a limit on how many clients can be using it at the same time. AVA can hit those limits as it runs multiple processes, but well-written services should emit an error or throttle in those cases. If the one you're using doesn't, the tests will hang.
+
+Use the `concurrency` flag to limit the number of processes ran. For example, if your service plan allows 5 clients, you should run AVA with `concurrency=5` or less.
+
+## Async operations
```
Comment (sindresorhus) [inline docs/common-pitfalls.md]: I opened an issue about this: https://github.com/avajs/eslint-plugin-ava/issues/122

We should really try to detect this at runtime and in our ESLint plugin too.

**Comment by sotojuan** [comment]
Fixed all the docs stuff—sorry about the unrelated commits (still not good at merge conflicts...), how would I fix that? Rebase?

**Comment by sindresorhus** [comment]
@sotojuan Doesn't matter. We squash on merge, so it's not included. You still have to fix the indentation changes in the readme though; https://github.com/avajs/ava/pull/919/files#diff-0730bb7c2e8f9ea2438b52e419dd86c9R163

**Comment by sotojuan** [comment]
All right, sorry about that, I did that commit in my work computer and probably didn't have editorconfig. Should be better now (?).

**Comment by sindresorhus** [comment]
Landing this now. We can fill in more as pitfalls are discovered. Awesome work @sotojuan :)
